x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/datafeed/extractor/aggregation/RollupDataExtractorFactory.java

```diff
@@ -0,0 +1,217 @@
+        return new RollupDataExtractor(client, dataExtractorContext);
+    }
+
+    public static void create(Client client,
+                              DatafeedConfig datafeed,
+                              Job job,
+                              Map<String, RollableIndexCaps> rollupJobsWithCaps,
+                              ActionListener<DataExtractorFactory> listener) {
+
+        final AggregationBuilder datafeedHistogramAggregation = getHistogramAggregation(
+            datafeed.getAggregations().getAggregatorFactories());
+        if ((datafeedHistogramAggregation instanceof DateHistogramAggregationBuilder) == false) {
+            listener.onFailure(
+                new IllegalArgumentException("Rollup requires that the datafeed configuration use a [date_histogram] aggregation," +
+                    " not a [histogram] aggregation over the time field."));
+            return;
+        }
+
+        final String timeField = ((ValuesSourceAggregationBuilder) datafeedHistogramAggregation).field();
+
+        Set<ParsedRollupCaps> rollupCapsSet = rollupJobsWithCaps.values()
+            .stream()
+            .flatMap(rollableIndexCaps -> rollableIndexCaps.getJobCaps().stream())
+            .map(rollupJobCaps -> ParsedRollupCaps.fromJobFieldCaps(rollupJobCaps.getFieldCaps(), timeField))
+            .collect(Collectors.toSet());
+
+        final long datafeedInterval = getHistogramIntervalMillis(datafeedHistogramAggregation);
+
+        List<ParsedRollupCaps> validIntervalCaps = rollupCapsSet.stream()
+            .filter(rollupCaps -> validInterval(datafeedInterval, rollupCaps))
+            .collect(Collectors.toList());
+
+        if (validIntervalCaps.isEmpty()) {
+            listener.onFailure(
+                new IllegalArgumentException(
+                    "Rollup capabilities do not have a [date_histogram] aggregation with an interval " +
+                        "that is a multiple of the datafeed's interval.")
+            );
+            return;
+        }
+        final List<ValuesSourceAggregationBuilder> flattenedAggs = new ArrayList<>();
+        flattenAggregations(datafeed.getAggregations().getAggregatorFactories(), datafeedHistogramAggregation, flattenedAggs);
+
+        if (validIntervalCaps.stream().noneMatch(rollupJobConfig -> hasAggregations(rollupJobConfig, flattenedAggs))) {
+            listener.onFailure(
+                new IllegalArgumentException("Rollup capabilities do not support all the datafeed aggregations at the desired interval.")
+            );
+            return;
+        }
+
+        listener.onResponse(new RollupDataExtractorFactory(client, datafeed, job));
+    }
+
+    private static boolean validInterval(long datafeedInterval, ParsedRollupCaps rollupJobGroupConfig) {
+        if (rollupJobGroupConfig.hasDatehistogram() == false) {
+            return false;
+        }
+        if ("UTC".equalsIgnoreCase(rollupJobGroupConfig.getTimezone()) == false) {
```

Comment:
out of curiosity: is that something for a follow-up? If the bucket span is compatible with shifting, it should not be a big deal to support other timezones?

Reply:
++

I imagine it only really matters for bucket spans of more than 30 minutes.  For ML 1 day bucket spans are always 24 hours and UTC, whereas for a date histogram with 1 day buckets (a) they could be offset from the UTC buckets the analytics expects to see and (b) if there is daylight saving then one bucket per year will be 23 hours and one will be 25 hours.  Also, for 1 hour bucket spans the offset problem could occur if the timezone is N.5 hours different from UTC, e.g. India at 5.5 hours ahead.  But for bucket spans up to and including 30 minutes we should be able to cope with any rollup timezone.

Or am I missing some other problem here?

Reply:
I think your description is correct @droberts195. We have gone through this before and the restriction of UTC timezone is already in place for aggregations (without rollups). Ben has just copied the validation over to the rollup realm. Since things might go wrong when dealing with non-UTC, imposing the UTC restriction is a good solution. It only requires the user to NOT change the default timezone of the date histogram aggregation. It does not pose any requirements on the actual data and how they're indexed.

Reply:
OK cool.  Let's leave any changes in this area for a separate PR then.  Definitely don't make any changes for this in this PR.

Reply:
Ah, just realised my comment above is confused. This is checking the rollup was in UTC. It is worth finding out what happens if you search a rollup which uses timezone X by aggregating over timezone Y. We need that before making a call on how we deal with this.